### ISSUE TITLE
[Bug] JSON_ARRAY Casts JSON_OBJECT as String, Causing JSON_EXTRACT to Fail

### ISSUE THREAD

**Issue by Baymine** [apache/doris#44565]
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

### Version

doris-2.1.6-rc04

### What's Wrong?

When using `JSON_ARRAY` in combination with `JSON_OBJECT`, the `JSON_OBJECT` appears to be cast as a string. This behavior prevents `JSON_EXTRACT` from retrieving the desired JSON keys correctly. However, in MySQL version 5.7.19, the SQL queries operate as expected without this issue.

### What You Expected?

The `JSON_EXTRACT` function should correctly retrieve the value associated with the specified JSON key without returning `<null>`. Specifically, extracting `'$[0].name'` from the JSON array should return `"Adam"`.

### How to Reproduce?

1. **Create JSON_OBJECT:**
   ```sql
   SELECT JSON_OBJECT("name", "Adam");
   ```
   **Output:**
   ```
   +-------------------------------+
   | json_object('name', 'Adam')   |
   +-------------------------------+
   | {"name":"Adam"}               |
   +-------------------------------+
   1 row in set (0.015s)
   ```

2. **Create JSON_ARRAY with JSON_OBJECT:**
   ```sql
   SELECT JSON_ARRAY(JSON_OBJECT("name", "Adam"));
   ```
   **Output:**
   ```
   +----------------------------------------------------+
   | json_array(json_object('name', 'Adam'), '66')      |
   +----------------------------------------------------+
   | ["{\"name\":\"Adam\"}"]                            |
   +----------------------------------------------------+
   1 row in set
   ```

3. **Attempt to Extract JSON Key:**
   ```sql
   SELECT JSON_EXTRACT(JSON_ARRAY(JSON_OBJECT("name", "Adam"), JSON_OBJECT("name", 'Kent')), '$.[0].name');
   ```
   **Output:**
   ```
   +--------------------------------------------------------------------------------------------------------------------+
   | json_extract(json_array(json_object('name', 'Adam'), json_object('name', 'Kent'), '66'), '$.[0].name')              |
   +--------------------------------------------------------------------------------------------------------------------+
   | <null>                                                                                                             |
   +--------------------------------------------------------------------------------------------------------------------+
   1 row in set (0.015s)
   ```

4. **MySQL 5.7.19 Behavior:**
   ```sql
   SELECT JSON_EXTRACT(JSON_ARRAY(JSON_OBJECT("name", "Adam"), JSON_OBJECT("name", 'Kent')), '$[0].name');
   ```
   **Output:**
   ```
   +-------------------------------------------------------------------------------------------------+
   | JSON_EXTRACT(JSON_ARRAY(JSON_OBJECT('name', 'Adam'), JSON_OBJECT('name', 'Kent')), '$[0].name') |
   +-------------------------------------------------------------------------------------------------+
   | "Adam"                                                                                          |
   +-------------------------------------------------------------------------------------------------+
   1 row in set (0.004s)
   ```

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

**Comment by ixzc** [comment]
yes, this is a bug. if you willing to fix this. you can submit a pr. or we can fix this.

**Comment by Baymine** [comment]
> yes, this is a bug. if you willing to fix this. you can submit a pr. or we can fix this.

Okay, I'll handle it

**Comment by TracebaK** [comment]
Hi @ixzc , is there anyone still working on a fix of this issue? I met a similar problem when I want to build a json string of a nested json object and pass the json string to my UDF. I used `json_object()` to build the json object and the array created using `json_array()` is converted to a string, which is not expected, like in the following example: 

```
my_udf(json_object('text', json_array('this is sentence 1', 'this is sentence2')))
``` 

What I got from my udf is:

```
{"text":"[\"this is sentence 1\",\"this is sentence2\"]"}
```

What I expected is something like: 

```
{"text":["this is sentence 1","this is sentence2"]}
```

**Comment by Timekeeper0327** [comment]
> Hi [@ixzc](https://github.com/ixzc) , is there anyone still working on a fix of this issue? I met a similar problem when I want to build a json string of a nested json object and pass the json string to my UDF. I used to build the json object and the array created using is converted to a string, which is not expected, like in the following example:`json_object()``json_array()`
> 
> ```
> my_udf(json_object('text', json_array('this is sentence 1', 'this is sentence2')))
> ```
> 
> What I got from my udf is:
> 
> ```
> {"text":"[\"this is sentence 1\",\"this is sentence2\"]"}
> ```
> 
> What I expected is something like:
> 
> ```
> {"text":["this is sentence 1","this is sentence2"]}
> ```

I’m having the same problem too. Could you please let me know if there’s a way to fix it?

**Comment by Baymine** [comment]
> > Hi [@ixzc](https://github.com/ixzc) , is there anyone still working on a fix of this issue? I met a similar problem when I want to build a json string of a nested json object and pass the json string to my UDF. I used to build the json object and the array created using is converted to a string, which is not expected, like in the following example:`` json_object()``json_array() ``
> > ```
> > my_udf(json_object('text', json_array('this is sentence 1', 'this is sentence2')))
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > What I got from my udf is:
> > ```
> > {"text":"[\"this is sentence 1\",\"this is sentence2\"]"}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > What I expected is something like:
> > ```
> > {"text":["this is sentence 1","this is sentence2"]}
> > ```
> 
> I’m having the same problem too. Could you please let me know if there’s a way to fix it?

I have proposed a fix:  https://github.com/apache/doris/pull/44897/files

**Comment by Timekeeper0327** [comment]
> > > Hi [@ixzc](https://github.com/ixzc) , is there anyone still working on a fix of this issue? I met a similar problem when I want to build a json string of a nested json object and pass the json string to my UDF. I used to build the json object and the array created using is converted to a string, which is not expected, like in the following example:` json_object()`json_array() ``
> > > ```
> > > my_udf(json_object('text', json_array('this is sentence 1', 'this is sentence2')))
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > What I got from my udf is:
> > > ```
> > > {"text":"[\"this is sentence 1\",\"this is sentence2\"]"}
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > What I expected is something like:
> > > ```
> > > {"text":["this is sentence 1","this is sentence2"]}
> > > ```
> > 
> > 
> > I’m having the same problem too. Could you please let me know if there’s a way to fix it?
> 
> I have proposed a fix: https://github.com/apache/doris/pull/44897/files

Thanks a lot for chiming in! Your suggestion makes perfect sense. I hadn’t considered that approach before, and I’m excited to test it out.
